index.js

```diff
@@ -78,64 +76,35 @@ module.exports = function combine () {
   N.each(allNeeds, function (_, path) {
     if (!N.get(allGives, path)) { throw new Error('export needed but not given' + path.join('.') + ' in: ' + path) }
   })
+}
 
-//  console.log("NEEDS", allNeeds)
-//  console.log("GIVES", allGives)
-
-  // okay, instead of iterating over everything, in dependency order.
-  // just create things in order added?
-
-  var sockets = {}
-  while (true) {
-    var newSockets = {}
-    // filter modules that have not been resolved yet.
-    modules = filter(modules, function (module, key) {
-      // if this module cannot be resolved yet, keep it for next time.
-      // collect the functions that this module needs.
-      var m = N.map(module.needs, function (type, path) {
-        var a = N.get(sockets, path)
-        if (!a) {
-          a = N.set(sockets, path, [])
-        }
-        return apply[type](a)
-      })
-
-      if(!isFunction(module.create)){throw new Error('did not have a create function', key)}
-      // create module, and get function(s) it returns.
-      var exported = module.create(m)
-      if(!exported) {throw new Error('export declared but not returned for', key)}
-
-      // for the functions it declares, merge these into newSockets
-      if (isString(module.gives)) {
-        append(newSockets, [module.gives], exported)
-      } else {
-        N.each(module.gives, function (_, path) {
-          var fun = N.get(exported, path)
-          append(newSockets, path, fun)
-        })
-      }
+function addExportsToApi (exported, api, module) {
+  if (isString(module.gives)) {
+    append(api, [module.gives], exported)
+  } else {
+    N.each(module.gives, function (_, path) {
+      var fun = N.get(exported, path)
+      append(api, path, fun)
     })
+  }
+}
 
-    if (isEmpty(newSockets)) {
-      throw new Error('could not resolve any modules')
-    } else {
-      N.each(newSockets, function (_ary, path) {
-        var ary = N.get(sockets, path) || N.set(sockets, path, [])
-        _ary.forEach(function (e) { ary.push(e) })
-      })
+function buildDeps (needs, api) {
+  return N.map(needs, function (type, path) {
+    var a = N.get(api, path)
+    if (!a) {
+      a = N.set(api, path, [])
     }
```

Comment:
this get || set pattern is repeated in append.
probably not worth doing anything about. I think the formatting on append was cleaner